GH-42156: [Java] Handle offset field from ArrowArray when BufferImportTypeVisitor imports offset buffer

vibhatha opened the pull request:
### Rationale for this change

This PR handles the usage of `offset` in `ArrowArray`. This is currently not used in Arrow Java implementation.

### What changes are included in this PR?

This PR uses the `ArrowArray` `offset` to calculate the offsets correctly in case of a sliced Array being provided. 
Corresponding test cases have been added. 

### Are these changes tested?

Tested using existing test cases and new test cases for slicing. 

### Are there any user-facing changes?

No
* GitHub Issue: apache/arrow-java#74
* GitHub Issue: #74

lidavidm reviewed (changes requested):
I'd expect there to be Java tests too? Even if you can't construct such an array normally in Java I expect you could construct one by hand instead

Also we need to test offset for all types

vibhatha commented:
> I'd expect there to be Java tests too? Even if you can't construct such an array normally in Java I expect you could construct one by hand instead

Sure, I can work on one. 

> 
> Also we need to test offset for all types

I guess we can do the similar slice test for all vector types with offsets using Python script. And probably we require equivalent Java tests using constructed array?

lidavidm commented:
Yes

lidavidm commented:
In general I'm surprised none of the existing integration tests use offsets...that is a wider problem with C Data

vibhatha commented:
> In general I'm surprised none of the existing integration tests use offsets...that is a wider problem with C Data

None of the ones in except for the one I just added doesn't use it. I will try to cover as much as ground possible.

lidavidm commented:
You may want to consult with maintainers from the other implementations too?

vibhatha commented on java/c/src/test/java/org/apache/arrow/c/RoundtripTest.java:
```diff
@@ -951,6 +952,86 @@ public void testImportReleasedArray() {
     }
   }
 
+  @Test
```
@lidavidm would this test be satisfactory to make sure we make the sliced vector manually in Java and test C data interface?

lidavidm replied:
Is this right? A sliced vector should be exported with the same buffers as a normal vector, but should have an offset applied. The actual buffers should not be changed

vibhatha replied:
Right, one question related to that. From C++ we could get the full offset, validity and data buffers, that's right. In Java, we would use the `arrowArrayOffset` to determine the first offset, and take dataLength + 1 from the first offset. This is a must right? 

And for validity buffer we can just import what is sent, and for data buffer, we need to determine the capacity, because the known information are via the adjustedOffset buffer. So that last value would have the length of the buffer we need to read. Though this will contain unnecessary values at the beginning of the buffer, but since we have correct offsets, the vector can interpret the data accurately. 

Is this understanding correct? Or does it have issues?

lidavidm replied:
Vector doesn't understand offsets. So all buffers have to be sliced. Also, validity buffer and offsets buffer need to be copied, because (1) you can't slice on a bit boundary and (2) the offsets need to be fixed

vibhatha replied:
Right. I see your point. This means we need to add the offset to the `ArrowArray.Snapshot` rather than creating the vector like this. I see.

lidavidm replied:
I'm still not sure about the tests here, I think the way we should do it is by manually tweaking the Snapshot instance

vibhatha commented on java/c/src/main/java/org/apache/arrow/c/BufferImportTypeVisitor.java:
```diff
@@ -212,19 +215,21 @@ public List<ArrowBuf> visit(ArrowType.FloatingPoint type) {
-          type,
-          start,
-          end);
-      final int len = end - start;
```
Not sure if this line was needed here? Or Am I missing something?

vibhatha commented:
@github-actions crossbow submit -g java

lidavidm commented on java/c/src/test/python/integration_tests.py:
```diff
@@ -54,11 +54,11 @@ def setup_jvm():
-    #     "-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:5005"
-    # ]
+    jvm_args = [
+        "-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:5005"
+    ]
 
-    jpype.startJVM(jpype.getDefaultJVMPath(), "-Djava.class.path=" + jar_path, **kwargs)
+    jpype.startJVM(jpype.getDefaultJVMPath(), "-Djava.class.path=" + jar_path, *jvm_args, **kwargs)
```
I don't think you meant to leave this?

vibhatha replied:
Oh yes, sorry, debug trace.

lidavidm commented on java/c/src/main/java/org/apache/arrow/c/BufferImportTypeVisitor.java:
```diff
@@ -116,7 +119,7 @@ private ArrowBuf importFixedBytes(ArrowType type, int index, long bytesPerSlot)
 
   private ArrowBuf importOffsets(ArrowType type, long bytesPerSlot) {
-    final long capacity = bytesPerSlot * (fieldNode.getLength() + 1);
+    final long capacity = bytesPerSlot * (fieldNode.getLength() + arrowArrayOffset + 1);
```
Why are we adjusting capacity here?

vibhatha replied:
Right. This is not needed.

vibhatha replied:
Oh wait, this is needed because this fails the `integration_tests.py` 

```bash
======================================================================
ERROR: test_binary_slice_array (__main__.TestPythonIntegration)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "Data.java", line 319, in org.apache.arrow.c.Data.importVector
java.lang.java.lang.IndexOutOfBoundsException: java.lang.IndexOutOfBoundsException: index (24) must not be greater than size (16)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "Data.java", line 319, in org.apache.arrow.c.Data.importVector
Exception: Java Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "integration_tests.py", line 202, in test_binary_slice_array
    self.round_trip_array(lambda: sliced_array)
  File "integration_tests.py", line 160, in round_trip_array
    self.bridge.python_to_java_array(original_arr, dictionary_provider) as vector:
  File "integration_tests.py", line 111, in python_to_java_array
    return self.java_c.Data.importVector(self.java_allocator, c_array, c_schema, dictionary_provider)
java.lang.java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: Could not load buffers for field : Binary. error message: index (24) must not be greater than size (16)

======================================================================
ERROR: test_string_slice_array (__main__.TestPythonIntegration)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "Data.java", line 319, in org.apache.arrow.c.Data.importVector
java.lang.java.lang.IndexOutOfBoundsException: java.lang.IndexOutOfBoundsException: index (24) must not be greater than size (16)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "Data.java", line 319, in org.apache.arrow.c.Data.importVector
Exception: Java Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "integration_tests.py", line 196, in test_string_slice_array
    self.round_trip_array(lambda: sliced_array)
  File "integration_tests.py", line 160, in round_trip_array
    self.bridge.python_to_java_array(original_arr, dictionary_provider) as vector:
  File "integration_tests.py", line 111, in python_to_java_array
    return self.java_c.Data.importVector(self.java_allocator, c_array, c_schema, dictionary_provider)
java.lang.java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: Could not load buffers for field : Utf8. error message: index (24) must not be greater than size (16)
```

vibhatha replied:
If we don't add the `arrowArrayOffset` then we get 4 slots, which are the first four slots, but what we need is not the first 4 slots, but the from `arrowArrayOffset` + 4 slots. And then we have to get the adjusted offset from this since Java doesn't need to keep the whole offset buffer like C++.  Am I missing something here?

lidavidm replied:
But we should be adjusting the start of the buffer, not the capacity?

lidavidm replied:
Or does the length have the offset pre-applied?

vibhatha replied:
`fieldNode.getLength()=3` as expected from slicing. Which means, we exactly don't know which 3 values to read here. I thought, since there is slicing, we need to read more from that buffer meaning, additional `arrowArrayOffset` amount. 

> But we should be adjusting the start of the buffer, not the capacity?

I completely agree, but I couldn't figure out a way to do that from the current code, we cannot provide a starting point, but a capacity to read from the beginning. Should we improve the memory read component?

lidavidm replied:
?? Just read the full buffer then slice.

vibhatha replied:
Right, in the C data interface `ArrowArray.Snapshot` only provides us `length` and the `offset`, other than that we have no information to determine the size of the full buffer. So what we read now has the capacity of `length + offset`. Am I missing something here?

lidavidm replied:
Right, but here I see only the full buffer, no slice

vibhatha replied:
Oh that's right. I will move the redundant logic in utf8 and binary type to here.

vibhatha commented on java/c/src/test/python/integration_tests.py:
Definitely, I just want to get an early feedback from you to see if the approach on the `VarCharVector` and the corresponding test case is okay or not? Then I am planning to mimic it to other types.

lidavidm commented on java/c/src/test/python/integration_tests.py:
```diff
@@ -190,6 +190,17 @@ def round_trip_reader(self, schema, batches):
         self.round_trip_array(lambda: pa.array([None, "a", "bb", "ccc"]))
 
+    def test_string_slice_array(self):
+        data = pa.array(["foo", "bar", "baz1", "baz223", "baz23445", "baz2121", "12312baz"])
```
Can we include some null and empty strings? (Both before and after the slice point)

lidavidm replied:
I'd also like to see a case where the sliced array is empty.

vibhatha replied:
Sure, let me add those cases.

lidavidm commented on java/c/src/test/java/org/apache/arrow/c/RoundtripTest.java:
```diff
@@ -951,6 +952,130 @@ public void testImportReleasedArray() {
+        // Producer exports vector into the C Data Interface structures
+        Data.exportVector(allocator, vector, null, arrowArray, arrowSchema);
+      }
+      // consumerArrowArray.snapshot().offset = offset;
```
this is commented out?

vibhatha replied:
I am still working on updating the test case as you recommended. Meaning, as you have mentioned again here `manually tweaking snapshot` is what we have to do. I am still working on it. I just pushed so that I can work from my other device.

vibhatha replied:
Sorry about the confusion. Still addressing reviews.

lidavidm commented on java/c/src/test/java/org/apache/arrow/c/RoundtripTest.java:
```diff
@@ -951,6 +952,130 @@ public void testImportReleasedArray() {
+      final int length = 3;
+      // create a sliced vector manually to mimic C++ slice behavior
+      VarCharVector slicedVector = (VarCharVector) getSlicedVector(vector, startIndex);
+      vector.splitAndTransferTo(startIndex, length, target);
```
I don't think this is actually a slice?

vibhatha commented:
@lidavidm I updated the test case for `VarCharVector`, but need to integrate the offset field for all other data types, which includes both functional changes and new test additions. I added the test skeleton, but may need some time to add the functions. I will ping once this is done.

vibhatha commented:
@lidavidm Thinking about the bit boundaries when it comes to setting up the validity buffer accurately. 
Would it be okay to reset the `validityBuffer` based on the offset? For all other buffers we can do the slicing, but for `validityBuffer` we cannot do that as we have that split in the bit boundary. Appreciate your thoughts.

lidavidm commented:
> @lidavidm Thinking about the bit boundaries when it comes to setting up the validity buffer accurately. Would it be okay to reset the `validityBuffer` based on the offset? For all other buffers we can do the slicing, but for `validityBuffer` we cannot do that as we have that split in the bit boundary. Appreciate your thoughts.

You generally have to copy when there is a slice, yes. (You can avoid it if offset % 8 == 0.)

vibhatha commented:
> You generally have to copy when there is a slice, yes. (You can avoid it if offset % 8 == 0.)

Right. So we will have a new buffer (new memory), how can we manage this memory which get created in the visitor?

lidavidm commented:
It will be owned by the vector?

lidavidm commented on .github/workflows/java.yml:
Did you mean to include this?

vibhatha replied:
@lidavidm I was testing Java log enabling via this PR. And this should be removed. Plus I haven't completed the required offset changes to all vector types yet.

lidavidm replied:
Can you move this to draft if it's not ready for review?

kou commented:
Could you open a new PR in apache/arrow-java?
